### PR TITLE
feat(vdsnapshot): add the new controller for the virtual disk snapshot

### PR DESCRIPTION
**Pull request by Isteb4k**
## Description

1. VirtualDiskSnapshot controller with two conditions: `VirtualDiskSnapshotReady` and `VirtualDiskReady`
2. Support for a new datasource for the disk: ObjectRef, kind - VirtualDiskSnapshot
3. condition `FilesystemReady` for the virtual machine
4. condition `Snapshotting` for the disk

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/virtualization-artifact/pkg/controller/service/snapshot_service.go [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+	path := fmt.Sprintf(virtualizationSubresourcePath, namespace, name, "freeze")
+
+	body, err := json.Marshal(&virtv1.FreezeUnfreezeTimeout{
+		UnfreezeTimeout: &metav1.Duration{},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = s.restClient.Put().AbsPath(path).Body(body).Do(ctx).Error()
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/service/snapshot_service.go:80]: I think, you should use our kubeclient.  
Add new method Freeze for VirtualMachineInterface. For example: https://github.com/deckhouse/virtualization/blob/1f2f3807ae8ec49e0133da064b1e17aa7536da71/api/client/kubeclient/vm.go#L99

File: images/virtualization-artifact/pkg/controller/service/snapshot_service.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+	}
+
+	return true, nil
+}
+
+func (s *SnapshotService) Unfreeze(ctx context.Context, name, namespace string) error {
+	path := fmt.Sprintf(virtualizationSubresourcePath, namespace, name, "unfreeze")
+
+	err := s.restClient.Put().AbsPath(path).Do(ctx).Error()
+	if err != nil {
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/service/snapshot_service.go:131]: Add new method Unfreeze for VirtualMachineInterface.

File: images/virtualization-artifact/pkg/controller/vd/internal/watcher/vdsnapshot_watcher.go [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+		Namespace: vdSnapshot.Namespace,
+	})
+	if err != nil {
+		slog.Default().Error(fmt.Sprintf("failed to list virtual disks: %s", err))
+		return
+	}
+
+	for _, vd := range vds.Items {
+		// Need to reconcile the virtual disk from which the snapshot was taken.
+		if vd.Name == vdSnapshot.Spec.VirtualDiskName {
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vd/internal/watcher/vdsnapshot_watcher.go:77]: Do we need to add indexes? For example:
https://github.com/deckhouse/virtualization/blob/1f2f3807ae8ec49e0133da064b1e17aa7536da71/images/virtualization-artifact/pkg/controller/indexer/indexer.go#L64

File: images/virtualization-artifact/pkg/controller/vm/internal/filesystem.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	if kvvmi == nil {
+		cb.Status(metav1.ConditionFalse).
+			Reason2(vmcondition.ReasonFilesystemNotReady).
+			Message(fmt.Sprintf("The internal virtual machine %s is not running.", changed.Name))
+		return reconcile.Result{}, nil
+	}
+
+	agentReady, _ := service.GetCondition(vmcondition.TypeAgentReady.String(), changed.Status.Conditions)
+	if agentReady.Status != metav1.ConditionTrue {
+		cb.Status(metav1.ConditionUnknown)
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/internal/filesystem.go:77]: Why Unknown?

File: images/virtualization-artifact/pkg/controller/vd/internal/snapshotting.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+	if vd.DeletionTimestamp != nil {
+		condition.Status = metav1.ConditionUnknown
+		condition.Reason = ""
+		condition.Message = ""
+		return reconcile.Result{}, nil
+	}
+
+	readyCondition, ok := service.GetCondition(vdcondition.ReadyType, vd.Status.Conditions)
+	if !ok || readyCondition.Status != metav1.ConditionTrue {
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vd/internal/snapshotting.go:59]: Why do you check  `readyCondition.Status != metav1.ConditionTrue` ?

File: images/virtualization-artifact/pkg/controller/vd/internal/snapshotting.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		condition.Status = metav1.ConditionUnknown
+		condition.Reason = ""
+		condition.Message = ""
+		return reconcile.Result{}, nil
+	}
+
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vd/internal/snapshotting.go:65]: need return this (if all conditions set in Unknown)

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: crds/doc-ru-virtualdisksnapshot.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
+spec:
+  versions:
+    - name: v1alpha2
+      schema:
+        openAPIV3Schema:
+          description: |
+            Предоставляет ресурс для создания снимков существующих виртуальных дисков, которые могут быть использованы в качестве источников данных для создания новых виртуальных дисков.
+
+            Под капотом автоматически создается ресурс `VolumeSnapshot`.
```
Comment (LopatinDmitr) [inline crds/doc-ru-virtualdisksnapshot.yaml:9]: I would suggest: "В процессе выполнения автоматически создается ресурс класса `VolumeSnapshot`."

The idea is to get rid of "Под капотом"

File: crds/virtualdisksnapshot.yaml [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                        type: string
+                      lastTransitionTime:
+                        description: Last time the condition transit from one status to another.
+                        format: date-time
+                        type: string
+                      message:
+                        description: Human readable message indicating details about last transition.
+                        type: string
+                      reason:
+                        description: (brief) reason for the condition's last transition.
```
Comment (LopatinDmitr) [inline crds/virtualdisksnapshot.yaml:82]: That is mean 'brief'?

File: images/virt-artifact/patches/022-debug.patch
Comment (LopatinDmitr) [inline images/virt-artifact/patches/022-debug.patch:1]: May be renaming this file?

File: crds/virtualdisksnapshot.yaml (hunk 8 above)
Comment (Isteb4k) [reply crds/virtualdisksnapshot.yaml:82]: Looks like it is used in the meaning of 'short description for the condition'. This is a common description for CRDs in accordance with other resources

File: images/virt-artifact/patches/022-debug.patch
Comment (Isteb4k) [reply images/virt-artifact/patches/022-debug.patch:1]: It was not added by me. Perhaps by mistake.

File: crds/virtualdisksnapshot.yaml [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+              properties:
+                virtualDiskName:
+                  type: string
+                  description: |
+                    The name of the virtual disk to take snapshot.
+                volumeSnapshotClassName:
+                  type: string
+                  description: |
+                    The name of the volume snapshot class to use while snapshotting virtual disk.
+                allowPotentiallyInconsistent:
```
Comment (fl64) [inline crds/virtualdisksnapshot.yaml:50]: Please, replace thist param with `requiredConsistency` with reverce logic

File: crds/virtualdisksnapshot.yaml [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                  type: boolean
+                  default: false
+                  description: |
+                    Specify whether to allow the snapshotting of a virtual machine's disk that is attached to a running virtual machine without an agent, and as a result, cannot be frozen.
+
+                    If value is false, the snapshot of the virtual disk will be taken only in the following scenarios:
+                    - the virtual disk is not attached to any virtual machine.
+                    - the virtual disk is attached to a virtual machine that is powered off.
+                    - the virtual disk is attached to a virtual machine with an agent, and the freeze operation was successful.
+            status:
```
Comment (fl64) [inline crds/virtualdisksnapshot.yaml:60]: There should be `consistent` param in status
with values: 
true - snapshot is consistent
nil - in other cases

File: crds/virtualdisksnapshot.yaml [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+                  enum:
+                    ["Pending", "InProgress", "Ready", "Failed", "Terminating"]
+                observedGeneration:
+                  type: integer
+                  description: |
+                    The generation last processed by the controller.
+      additionalPrinterColumns:
+        - name: Phase
+          type: string
+          jsonPath: .status.phase
```
Comment (fl64) [inline crds/virtualdisksnapshot.yaml:117]:
```suggestion
      additionalPrinterColumns:
        - name: Phase
          type: string
          jsonPath: .status.phase
        - name: Age
          type: date
          jsonPath: .metadata.creationTimestamp
```

File: crds/virtualdisksnapshot.yaml [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+                    * Ready - the snapshot creation has successfully completed, and the virtual disk snapshot is now available.
+                    * Failed - an error occurred during the snapshotting process.
+                    * Terminating - the resource is in the process of being deleted.
+                  enum:
+                    ["Pending", "InProgress", "Ready", "Failed", "Terminating"]
+                observedGeneration:
+                  type: integer
+                  description: |
+                    The generation last processed by the controller.
+      additionalPrinterColumns:
```
Comment (fl64) [inline crds/virtualdisksnapshot.yaml:114]: Mb add a field with information about whether the snapshot is consistent?

**Review by fl64 — APPROVED** [review]

**Review by yaroslavborbat — DISMISSED** [review]

**Review by LopatinDmitr — APPROVED** [review]

**Pull request merged**